server/serializers.py

```diff
@@ -265,10 +267,15 @@ def jsonize_clientside_tab(tab: clientside.TabUpdate) -> Dict[str, Any]:
 
 
 def jsonize_i18n_message(message: I18nMessage, ctx: JsonizeContext) -> str:
+    assert message.source is None
     if message.id == "TODO_i18n":
         return message.args["text"]
     else:
-        raise RuntimeError("TODO_i18n")
+        return localize(
```

Comment:
What triggers the `or`? I think we need a comment here.

I feel bad asking you to do something I myself neglected to do, but ... well, I will :). Could you please create `server/tests/test_serializers.py` and unit-test the three code paths here? (`TODO_i18n`; `localize(ctx.locale_id, ...)`; `not localize(ctx.locale_id, ...)`)

Reply:
The intention is this procedure
- localize in given locale, if there is a message for it
- localize in default locale, if there is a message for it
- return message id, which is better than nothing; we could also throw an exception here if you like

I will add a comment and unit tests

Reply:
Ohhh, I had misunderstood the `localize()` function's behavior. Let's take our time here....

First: could you please tweak the docs in `i18n/trans.py` to clarify this? The `localize()` method should:

* Be type-annotated. It returns `Optional[str]` -- that's not obvious at all.
* Explicitly state: `Return None if there is no translation in the given locale and no default.`
* While we're at it, explicitly state: `Raise ICUError if the default message is invalid (programmer error).`

Ditto for `trans()`.

My instinct tells me we're going about this the wrong way here -- this double-invocation of `localize()` feels different from our `trans()` calls. We could make them feel the same by making `MessageTranslator` use the default _catalog_ instead of the default _message_. (`MessageTranslator.trans()` would not take a `default` argument.) Does that sound tempting?

Reply:
(Or another idea: `localize()` could look up `default` itself from the default catalog and pass it to `MessageTranslator.trans()`. That might give the same outcome with fewer code changes.)

Reply:
The ICU locale stored in `MessageTranslator` should match the locale of the catalog stored in `MessageTranslator`. If it does not, there will be inconsistent behaviour; for example when asking for a Ukranian text ([they use](http://cldr.unicode.org/index/cldr-spec/plural-rules) the `one` rule also for some values ending in 1) and an English message comes up from the default catalog, the case for `one` will be used when the value is 21, 31, etc. Moreover, `MessageTranslator` already has many responsibilities (fetch message from catalog, replace parameters, and replace tags if needed).

We could move the multiple invocations within `trans`, `localize`, and `localize_html`: they could get the `MessageTranslator` for their given locale, fall back to the `MessageTranslator` for the default locale, and then fall back to `default`; they could even ignore (`trans`) or not have (`localize` and `localize_html`) the `default` argument. Without `default`, they could return `None` or the `message_id`.

Reply:
Your concern about `MessageTranslator` makes me realize ... we already have a bug here, right? When `MessageTranslator.trans()` translates the fallback, it does it in the catalog locale, _not_ English.

It would certainly feel comfortable to me if `localize()` didn't have a `default` argument -- and if the `trans()` implementation simply called `localize()`. If you like that approach, too, please go for it.

All of this suggests to me a direction that would clarify the `localize()` API and simplify `MessageTranslator` a bit:

* Nix `default` from `MessageTranslator.trans()`, and move the fallback logic into `localize()` itself.
* Change the module-level `trans()` to invoke `localize()`

For this pull request, I'm only _really_ concerned with the `localize()` _API_. (These fallbacks are a low priority since they're never meant to show up.) But I'm curious if you agree that A) the ICU-locale bug you describe already exists; and B) my suggestion would address it neatly.

Reply:
Oops! Indeed, this bug already exists in `MessageTranslator`! Your suggestion looks great.

Reply:
I refactored and added a few more unit tests. Now `trans` and `jsonize_i18n_message` use `localize`, while `trans_html` (template tag) uses `localize_html`.  `localize` and `localize_html` do the "try this locale, else try default locale, else return None" stuff and `MessageTranslator` is not concerned with defaults at all. It looks way better.

I opted for everything returning `None` when no messages have been found. This will prompt developers to run `bin/dev extract-messages` earlier, since they will not see their texts otherwise. Keep in mind that it would also make some sense to make the upper-level functions, i.e. `trans`, `trans_html` (template tag) and `jsonize_18n_message` return the message id instead of `None`; we can decide it later.